# Pytorch Basics

## [1] Pytorch Tensor Basic Usage

- Create Tensor
- Joining, Slicing
- Math Operations
- Initialization

## 1. Create Tensor

### 1) random numbers

In [1]:
import torch

# Pytorch Version 확인
print(torch.__version__)

1.0.0


In [2]:
# torch.rand(size)
# 0~1 사이의 연속균등분포에서 값을 뽑아 5x3 텐서를 생성합니다. ~U(0,1)
x = torch.rand(2,3)
x

tensor([[0.0558, 0.3173, 0.0537],
        [0.4884, 0.8494, 0.1010]])

In [3]:
# torch.randn(size)
# randn의 경우 평균은 0, 표준편차는 1인 정규분포에서 값을 가져옵니다. ~z(0,1)
x = torch.randn(2,3)
x

tensor([[ 1.2258, -0.6182, -0.5214],
        [-0.6280, -0.6622, -0.8422]])

### 2) empty, zeros, ones, arange, max

In [4]:
# torch.empty(size)
# 원하는 크기의 빈 텐서를 생성합니다.
x = torch.empty(2,3)
x

tensor([[1.4013e-45, 0.0000e+00, 1.4013e-45],
        [0.0000e+00, 0.0000e+00, 0.0000e+00]])

In [5]:
# torch.zeros(size)
# 원하는 크기의 빈 텐서를 생성합니다.
x = torch.zeros(2,3)
x

tensor([[0., 0., 0.],
        [0., 0., 0.]])

In [6]:
# torch.ones(size)
# 원하는 크기의 1의 값을 가진 텐서를 생성합니다.
x = torch.ones(2,3)
x

tensor([[1., 1., 1.],
        [1., 1., 1.]])

In [7]:
# torch.max(input, dim, keepdim=False, out=None) -> (Tensor, LongTensor)
# tensor의 최대값을 산출합니다.

x = torch.rand(2,3)
print(x)

# 최대값을 출력합니다.
print(torch.max(x))

# 해당 차원에서 최대값과 그 위치를 출력합니다.
print(torch.max(x,1))

# 위치를 출력합니다.
print(torch.max(x,1)[1])

tensor([[0.6047, 0.7987, 0.6312],
        [0.6695, 0.7946, 0.7557]])
tensor(0.7987)
(tensor([0.7987, 0.7946]), tensor([1, 1]))
tensor([1, 1])


### 3) Tensor Data Type

| Data type               | dtype                                         | Tensor types                 |
|:-------------------------|:-----------------------------------------------|:------------------------------|
| 32-bit floating point   | ``` torch.float32 ``` or ``` torch.float ```  | ``` torch.*.FloatTensor ```  |
| 64-bit floating point   | ``` torch.float64 ``` or ``` torch.double ``` | ``` torch.*.DoubleTensor ``` |
| 16-bit floating point   | ``` torch.float16 ``` or ``` torch.half ```   | ``` torch.*.HalfTensor ```   |
| 16-bit integer (signed) | ``` torch.int16 ``` or ``` torch.short ```    | ``` torch.*.ShortTensor ```  |
| 32-bit integer (signed) | ``` torch.int32 ``` or ``` torch.int ```      | ``` torch.*.IntTensor ```    |
| 64-bit integer (signed) | ``` torch.int64 ``` or ``` torch.long ```     | ``` torch.*.LongTensor ```   |

In [8]:
# torch.empty(size, dtype=)
# torch.zeros(size, dtype=)
# torch.arange(start,end,dtype=)

# torch에서 지원하는 다양한 data type을 옵션으로 지정할 수 있습니다.

x = torch.empty(2,3, dtype=torch.float)
y = torch.zeros(2,3, dtype = torch.long)
z = torch.arange(0,3,step=0.5,dtype=torch.double)

print(x)
print(y)
print(z)

tensor([[-3.9982e-07,  4.5911e-41,  1.4013e-45],
        [ 0.0000e+00,  1.4013e-45,  0.0000e+00]])
tensor([[0, 0, 0],
        [0, 0, 0]])
tensor([0.0000, 0.5000, 1.0000, 1.5000, 2.0000, 2.5000], dtype=torch.float64)


In [9]:
# torch.tensor(size or list, dtype = torch.floattensor) 
# 원하는 텐서를 바로 생성합니다.
x = torch.tensor([5.5,3])
x

tensor([5.5000, 3.0000])

In [10]:
# torch.FloatTensor(size or list)
x = torch.FloatTensor(2,3)
print(x)
x = torch.FloatTensor([2,3])
print(x)

tensor([[-3.9982e-07,  4.5911e-41,  0.0000e+00],
        [ 0.0000e+00,  1.4013e-45,  0.0000e+00]])
tensor([2., 3.])


In [11]:
# tensor.type_as(tensor_type)
# tensor 형 변환
x = x.type_as(torch.IntTensor())
x

tensor([2, 3], dtype=torch.int32)

### 4) Tensor Size

In [12]:
# tensor.size()

x = torch.FloatTensor(10,12,3,3)

print(x.size()[:])
print(x.size()[0])

torch.Size([10, 12, 3, 3])
10


## 2. Math Operations

### 1) add, mul, div

In [13]:
# torch.add()
# 텐서의 값끼리 더합니다.

x1 = torch.FloatTensor([[1,2,3],[4,5,6]])
x2 = torch.FloatTensor([[1,2,3],[4,5,6]])
add = torch.add(x1, x2)

# torch.add는 +로 대체 가능
print(add)
print(x1+x2)
print(x1-x2)

tensor([[ 2.,  4.,  6.],
        [ 8., 10., 12.]])
tensor([[ 2.,  4.,  6.],
        [ 8., 10., 12.]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])


In [14]:
#차원이 다르더라도 안쪽 차원만 같으면 연산이 가능합니다.

x = torch.rand(1,5,3)
y = torch.rand(5,3)
torch.add(x,y)

tensor([[[1.2627, 0.3626, 1.0136],
         [0.8887, 1.3566, 1.0730],
         [1.2860, 1.5769, 1.1165],
         [1.3708, 1.0847, 0.3843],
         [0.8029, 0.1814, 0.9521]]])

In [15]:
# torch.mul()
# hadamard product, mxn, mxn 행렬 2개를 곱합니다. 덧셈에 대해 분배법칙을 따릅니다.

x1 = torch.FloatTensor([[1,2,3],[4,5,6]])
x2 = torch.FloatTensor([[1,2,3],[4,5,6]])
x3 = torch.mul(x1,x2)

# torch.mul은 *로 대체 가능
print(x3)
print(x1*x2)

tensor([[ 1.,  4.,  9.],
        [16., 25., 36.]])
tensor([[ 1.,  4.,  9.],
        [16., 25., 36.]])


In [16]:
# torch.div()

x1 = torch.FloatTensor([[1,2,3],[4,5,6]])
x2 = torch.FloatTensor([[1,2,3],[4,5,6]])
x3 = torch.div(x1,x2)

# torch.div는 /로 대체 가능
print(x3)
print(x1/x2)

tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])


### 2) pow, exp, log

In [17]:
# torch.pow(input,exponent)

x1 = torch.rand(3,4)

print(x1)
# torch.pow는 **로 대체 가능
print(torch.pow(x1,2),"\n",x1**2)

tensor([[0.5592, 0.9792, 0.1183, 0.4721],
        [0.3496, 0.0501, 0.0639, 0.7570],
        [0.0884, 0.2452, 0.2394, 0.2675]])
tensor([[0.3127, 0.9588, 0.0140, 0.2228],
        [0.1222, 0.0025, 0.0041, 0.5730],
        [0.0078, 0.0601, 0.0573, 0.0716]]) 
 tensor([[0.3127, 0.9588, 0.0140, 0.2228],
        [0.1222, 0.0025, 0.0041, 0.5730],
        [0.0078, 0.0601, 0.0573, 0.0716]])


In [18]:
# torch.exp(tensor,out=None)

x1 = torch.rand(3,4)

print(x1)
print(torch.exp(x1))

tensor([[0.6543, 0.0180, 0.3945, 0.4455],
        [0.3589, 0.1810, 0.3855, 0.4285],
        [0.3367, 0.5431, 0.5213, 0.1646]])
tensor([[1.9238, 1.0182, 1.4836, 1.5613],
        [1.4317, 1.1984, 1.4703, 1.5349],
        [1.4004, 1.7213, 1.6842, 1.1789]])


In [19]:
# torch.log(input,out=None)

x1 = torch.rand(3,4)

print(x1)
print(torch.log(x1))

tensor([[0.6589, 0.3993, 0.9101, 0.2046],
        [0.6485, 0.4819, 0.9604, 0.9543],
        [0.8220, 0.1312, 0.9094, 0.2397]])
tensor([[-0.4171, -0.9181, -0.0942, -1.5865],
        [-0.4330, -0.7300, -0.0404, -0.0468],
        [-0.1960, -2.0310, -0.0950, -1.4282]])


In [20]:
# torch.mm(matrix1, matrix2)

x1 = torch.rand(3,4)
x2 = torch.rand(4,5)

# torch.mul과 구분해서 사용해야함. mm은 matrix multiplication 연산을 수행
torch.mm(x1,x2)

tensor([[0.4302, 0.6682, 0.1026, 0.8251, 0.7992],
        [0.9914, 1.0428, 0.5859, 1.0643, 0.7807],
        [0.5313, 0.6795, 0.1847, 0.8280, 0.7615]])

In [21]:
# torch.bmm(batch_matrix1, batch_matrix2)

x1 = torch.rand(10,3,4)
x2 = torch.rand(10,4,5)

# Batch끼리의 계산을 할 때 자주 사용됨
torch.bmm(x1,x2).size()

torch.Size([10, 3, 5])

![img](https://github.com/shwksl101/Pytorch-A-to-Z/blob/master/img/innerproduct.PNG?raw=true)

In [22]:
# torch.dot(tensor1, tensor2)

torch.dot(torch.tensor([2,3]), torch.tensor([2,1]))

tensor(7)

In [23]:
# torch.t(matrix)
# tranpose

x1 = torch.rand(3,4)
print(x1.size(), x1.t().size())

# torch.transpose(input,dim0,dim1)

x1 = torch.rand(10,3,4)
x1.size(), torch.transpose(x1,1,2).size(), x1.transpose(1,2).size()

torch.Size([3, 4]) torch.Size([4, 3])


(torch.Size([10, 3, 4]), torch.Size([10, 4, 3]), torch.Size([10, 4, 3]))

### 3) view, item

In [24]:
# view()

x = torch.randn(4,4)
y = x.view(16)
z = x.view(-1,8) # -1의 경우 다른 차원들로 유추합니다. 이 경우에는 2로 유추합니다.
print(x.size())
print(y.size(),y)
print(z.size(),z)

torch.Size([4, 4])
torch.Size([16]) tensor([-1.3522,  1.3358,  0.9015,  1.1398, -0.0273,  0.0384,  0.7265,  0.1332,
         0.8172, -0.9961,  0.5111,  0.5698,  0.5933, -1.4054, -1.9977,  0.9106])
torch.Size([2, 8]) tensor([[-1.3522,  1.3358,  0.9015,  1.1398, -0.0273,  0.0384,  0.7265,  0.1332],
        [ 0.8172, -0.9961,  0.5111,  0.5698,  0.5933, -1.4054, -1.9977,  0.9106]])


In [25]:
# item()

x = torch.randn(1)
print(x)
print(x.item())

tensor([0.3563])
0.3562667965888977


- 더 많은 연산은 다음의 링크에서 참고하세요 https://pytorch.org/docs/stable/torch.html

## 3. Tensor to Numpy, Numpy to Tensor

In [26]:
import numpy as np

In [27]:
a = torch.ones(5) # 1로 채워진 텐서를 생성합니다.
print(a)

tensor([1., 1., 1., 1., 1.])


In [28]:
# tensor.numpy()

b = a.numpy()
print(b)

[1. 1. 1. 1. 1.]


In [29]:
# torch.from_numpy(ndarray)

a = np.ones(5) #a와 b는 연동됩니다.
b = torch.from_numpy(a)

np.add(a,1,out=a)

print(a)
print(b) #chartensor를 제외한 모든 tensor는 numpy로의 변환을 지원합니다.

[2. 2. 2. 2. 2.]
tensor([2., 2., 2., 2., 2.], dtype=torch.float64)


## 4. Tensor on GPU

단, GPU를 사용하기 위해서는 CUDA와 cudnn을 먼저 설치해야함.

In [30]:
x= torch.FloatTensor([[1,2,3],[4,5,6]])

if torch.cuda.is_available():
    x_cuda = x.cuda()
else:
    x_cuda = x.cpu()
    
print(x_cuda)

tensor([[1., 2., 3.],
        [4., 5., 6.]], device='cuda:0')


In [31]:
if torch.cuda.is_available():
    device = torch.device("cuda")
    y = torch.ones_like(x,device=device)
    x = x.to(device)
    z = x+y
    print(z)
    print(z.to("cpu",torch.double)) #.to 기능을 tensor를 GPU로 연산할 수 있습니다

tensor([[2., 3., 4.],
        [5., 6., 7.]], device='cuda:0')
tensor([[2., 3., 4.],
        [5., 6., 7.]], dtype=torch.float64)


## 5. Slicing, Joining

### 1) Joining

In [32]:
# torch.cat(seq, dim=0)
# dim을 기준으로 tensor를 합칩니다.
# dim = 0은 행, dim = 1은 열 기준입니다.

x = torch.FloatTensor([[1,2,3],
                       [4,5,6]])
y = torch.FloatTensor([[7,8,9],
                       [10,11,12]])
z1 = torch.cat([x,y],dim = 0)
z2 = torch.cat([x,y],dim = 1)

print(x)
print(y)
print(z1)
print(z2)

tensor([[1., 2., 3.],
        [4., 5., 6.]])
tensor([[ 7.,  8.,  9.],
        [10., 11., 12.]])
tensor([[ 1.,  2.,  3.],
        [ 4.,  5.,  6.],
        [ 7.,  8.,  9.],
        [10., 11., 12.]])
tensor([[ 1.,  2.,  3.,  7.,  8.,  9.],
        [ 4.,  5.,  6., 10., 11., 12.]])


In [33]:
# torch.stack(sequence,dim=0)
# dim을 기준으로 쌓습니다.

x = torch.FloatTensor([1,2,3])
x_stack = torch.stack([x,x],dim=0)
x_stack2 = torch.stack([x,x],dim=1)

print(x_stack,"\n",x_stack2)

tensor([[1., 2., 3.],
        [1., 2., 3.]]) 
 tensor([[1., 1.],
        [2., 2.],
        [3., 3.]])


### 2) Slicing

In [34]:
# torch.chunk(tensor, chunks, dim=0)
# tensor를 chunk 단위로 쪼갭니다.

x = torch.FloatTensor([[1,2,3],
                       [4,5,6]])
y = torch.FloatTensor([[7,8,9],
                       [10,11,12]])
z1 = torch.cat([x,y],dim = 0)

x_1, x_2 = torch.chunk(z1,2,dim=0)
y_1, y_2, y_3 = torch.chunk(z1,3,dim=1)

print(z1)
print(x_1)
print(x_2)
print(y_1)
print(y_2)
print(y_3)

tensor([[ 1.,  2.,  3.],
        [ 4.,  5.,  6.],
        [ 7.,  8.,  9.],
        [10., 11., 12.]])
tensor([[1., 2., 3.],
        [4., 5., 6.]])
tensor([[ 7.,  8.,  9.],
        [10., 11., 12.]])
tensor([[ 1.],
        [ 4.],
        [ 7.],
        [10.]])
tensor([[ 2.],
        [ 5.],
        [ 8.],
        [11.]])
tensor([[ 3.],
        [ 6.],
        [ 9.],
        [12.]])


In [35]:
# torch.splot(tensor,splot_size,dim=0)

x = torch.FloatTensor([[1,2,3],
                       [4,5,6]])
y = torch.FloatTensor([[7,8,9],
                       [10,11,12]])
z1 = torch.cat([x,y],dim = 0)

x1,x2 = torch.split(z1,2,dim=0)
y1 = torch.split(z1,2,dim=1)

print(x1)
print(x2)
print(y1)

tensor([[1., 2., 3.],
        [4., 5., 6.]])
tensor([[ 7.,  8.,  9.],
        [10., 11., 12.]])
(tensor([[ 1.,  2.],
        [ 4.,  5.],
        [ 7.,  8.],
        [10., 11.]]), tensor([[ 3.],
        [ 6.],
        [ 9.],
        [12.]]))


### 3) squeezing

In [36]:
# torch.squeez(input, dim=None)
# 1짜리 차원을 줄입니다.

x1 = torch.FloatTensor(10,1,3,1,4)
x2 = torch.squeeze(x1)

x1.size(), x2.size()

(torch.Size([10, 1, 3, 1, 4]), torch.Size([10, 3, 4]))

In [37]:
# torch.unsqueeze(input,dim=None)
# 1짜리 차원을 더합니다.

x1 = torch.FloatTensor(10,3,4)
x2 = torch.unsqueeze(x1,dim=0)

x.size(), x2.size()

(torch.Size([2, 3]), torch.Size([1, 10, 3, 4]))

## 6. Initialization

In [38]:
import torch.nn.init as init

x1 = init.uniform_(torch.FloatTensor(3,4),a=0,b=9) 
x2 = init.normal_(torch.FloatTensor(3,4),std=0.2)
x3 = init.constant_(torch.FloatTensor(3,4),3.1415)

x1,x2,x3

(tensor([[5.2884, 4.4637, 7.2311, 0.3083],
         [0.6230, 0.5624, 7.0414, 0.6258],
         [6.6145, 6.7477, 2.8904, 8.1059]]),
 tensor([[-0.0589, -0.0038, -0.0622, -0.0431],
         [ 0.0394, -0.0435,  0.1642, -0.0922],
         [ 0.1898, -0.1107, -0.1054, -0.4502]]),
 tensor([[3.1415, 3.1415, 3.1415, 3.1415],
         [3.1415, 3.1415, 3.1415, 3.1415],
         [3.1415, 3.1415, 3.1415, 3.1415]]))

## [2] Autograd

### Autograd Package

- Autograd 패키지는 tensor의 모든 연산에 자동 미분을 제공합니다. 이는 define-by-run의 프레임워크로 코드를 어떻게 작성하느냐에 따라 역전파가 정의된다는 뜻입니다. 역전파는 학습과정의 매 단계마다 달라집니다.

- .requires_grad 속성을 True로 설정하면 해당 tensor의 모든 연산을 추적합니다. 계산이 완료된 후 .backward()를 호출해 gradient를 자동으로 계산할 수 있습니다. 이 tensor의 gradient는 .grad에 누적됩니다.

- 연산 기록을 추적하는 것을 멈추기 위해 코드 블럭을 with torch.no_grad():로 감쌀 수 있습니다. gradient는 필요 없지만 requires_grad=True가 설정되어 학습 가능한 Parameter(매개변수)를 갖는 모델을 평가할 때 유용합니다.

In [39]:
x = torch.ones(2,2,requires_grad=True) #tensor를 생성하고 requires_grad=True로 연산을 기록합니다.
print(x)

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)


In [40]:
y = x+2 #gradient function이 자동으로 포함됩니다.
print(y)

tensor([[3., 3.],
        [3., 3.]], grad_fn=<AddBackward0>)


In [41]:
z = y*y*3
out = z.mean()
print(z,out)

tensor([[27., 27.],
        [27., 27.]], grad_fn=<MulBackward0>) tensor(27., grad_fn=<MeanBackward1>)


In [42]:
a = torch.randn(2,2)
a = ((a*3)/(a-1)) 
print(a.requires_grad)
print(a.grad_fn) # 사용자가 만든 텐서의 grad_fn은 none입니다.
a.requires_grad_(True)
print(a.requires_grad)
b = (a*a).sum()
print(b.grad_fn) #requires_grad_(True)로 지정하고 연산하면 이렇게 grad_fn가 생깁니다.

False
None
True


### Gradient

In [43]:
print(out) # out = 3(x+2)*2
out.backward()

tensor(27., grad_fn=<MeanBackward1>)


In [44]:
print(x)
print(x.grad) # d(out)/dx 를 출력합니다.

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)
tensor([[4.5000, 4.5000],
        [4.5000, 4.5000]])


In [45]:
x = torch.randn(3,requires_grad=True)

y=x*2

while y.data.norm() < 1000:
    
    #data.norm()은 점들 사이의 유클리디안 거리를 나타냅니다
    #torch.sqrt(torch.sum(torch.pow(y, 2)))
    
    y = y*2
    
    print(y,y.data.norm())

tensor([-4.0405,  2.6186, -9.0408], grad_fn=<MulBackward0>) tensor(10.2430)
tensor([ -8.0810,   5.2371, -18.0817], grad_fn=<MulBackward0>) tensor(20.4860)
tensor([-16.1620,  10.4743, -36.1634], grad_fn=<MulBackward0>) tensor(40.9721)
tensor([-32.3241,  20.9486, -72.3267], grad_fn=<MulBackward0>) tensor(81.9442)
tensor([ -64.6482,   41.8971, -144.6535], grad_fn=<MulBackward0>) tensor(163.8883)
tensor([-129.2963,   83.7943, -289.3070], grad_fn=<MulBackward0>) tensor(327.7767)
tensor([-258.5927,  167.5885, -578.6140], grad_fn=<MulBackward0>) tensor(655.5533)
tensor([ -517.1854,   335.1770, -1157.2279], grad_fn=<MulBackward0>) tensor(1311.1067)


In [46]:
gradients = torch.tensor([0.1,1.0,0.0001],dtype=torch.float)
print(y)
y.backward(gradients)
print(x.grad) # d(y)/d(x) 를 출력합니다

tensor([ -517.1854,   335.1770, -1157.2279], grad_fn=<MulBackward0>)
tensor([5.1200e+01, 5.1200e+02, 5.1200e-02])


In [47]:
print(x.requires_grad)
print((x**2).requires_grad)

with torch.no_grad():
    print((x**2).requires_grad) #tensor들의 연산 기록 추적을 막을 수 있습니다.
     
print((x**2).requires_grad)

True
True
False
True


- autograd package에 대한 더 자세한 정보는 다음의 링크를 참고하세요. https://pytorch.org/docs/stable/torch.html

## [3] Neural Network 기본 실습

In [48]:
import torch
import torch.nn as nn
import torch.optim as optim

from torch.autograd import Variable
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision import datasets

batch_size = 32
learning_rate = 0.01
num_epochs = 5

# MNIST Dataset DataLoader
train_dataset = datasets.MNIST(root='./data', train=True,
                        transform=transforms.ToTensor(), download=True)
test_dataset = datasets.MNIST(root='./data', train=False,
                        transform=transforms.ToTensor())

train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

In [49]:
# Neural Network Modeling
class Neuralnetwork(nn.Module):
    def __init__(self, num_classes=10):
        super(Neuralnetwork, self).__init__()
        self.layer1 = nn.Linear(28*28, 100)
        self.layer2 = nn.Linear(100, 200)
        self.layer3 = nn.Linear(200, num_classes)
        
    def forward(self, x):
        out = x.view(x.size(0), -1)
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        
        return out

In [50]:
# Model
model = Neuralnetwork()

# Loss
criterion = nn.CrossEntropyLoss()

# Optimizer
optimizer = optim.SGD(model.parameters(), lr=learning_rate)

In [51]:
# Traing the model
for epoch in range(num_epochs):
    for i, (img, label) in enumerate(train_loader, 1):
        img, label = Variable(img), Variable(label)
        out = model(img)
        loss = criterion(out, label)
        
        # Backward and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
        if (i+1) % 100 == 0:
            print('Epoch [{}/{}], Step [{}/{}], Loss: {:.4f}'.format(epoch+1, num_epochs, i+1, len(train_loader), loss.item()))

Epoch [1/5], Step [100/1875], Loss: 2.1738
Epoch [1/5], Step [200/1875], Loss: 1.8190
Epoch [1/5], Step [300/1875], Loss: 1.3110
Epoch [1/5], Step [400/1875], Loss: 1.0612
Epoch [1/5], Step [500/1875], Loss: 0.8907
Epoch [1/5], Step [600/1875], Loss: 0.6475
Epoch [1/5], Step [700/1875], Loss: 0.9325
Epoch [1/5], Step [800/1875], Loss: 0.5230
Epoch [1/5], Step [900/1875], Loss: 0.4543
Epoch [1/5], Step [1000/1875], Loss: 0.3658
Epoch [1/5], Step [1100/1875], Loss: 0.4579
Epoch [1/5], Step [1200/1875], Loss: 0.3607
Epoch [1/5], Step [1300/1875], Loss: 1.0041
Epoch [1/5], Step [1400/1875], Loss: 0.3750
Epoch [1/5], Step [1500/1875], Loss: 0.4524
Epoch [1/5], Step [1600/1875], Loss: 0.3276
Epoch [1/5], Step [1700/1875], Loss: 0.6145
Epoch [1/5], Step [1800/1875], Loss: 0.1755
Epoch [2/5], Step [100/1875], Loss: 0.3312
Epoch [2/5], Step [200/1875], Loss: 0.2156
Epoch [2/5], Step [300/1875], Loss: 0.2872
Epoch [2/5], Step [400/1875], Loss: 0.2551
Epoch [2/5], Step [500/1875], Loss: 0.5129
Ep

In [52]:
# Test the model
model.eval()

correct = 0
total = 0
for img, label in test_loader:
    out = model(img)
    _, predicted = torch.max(out.data, 1)
    total += label.size(0)
    correct += (predicted == label).sum().item()
    
print('Test Accuracy of the model on the 10000 test images: {} %'.format(100 * correct / total))

# Save the model checkpoint
torch.save(model.state_dict(), 'model_nn.ckpt')

Test Accuracy of the model on the 10000 test images: 91.99 %


### [과제] Neural Network의 구조를 변경하여, MNIST Classification Accuracy 92% 이상으로 만들기